Get kostal plenticore hostname id from get_settings

erikbadman opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #89949
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

erikbadman committed:
Get hostname id from get_settings

stegm commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -32,9 +32,17 @@ async def test_connection(hass: HomeAssistant, data) -> str:
 
-    return values["scb:network"]["Hostname"]
+        # get name of the hostname setting id
+        all_settings = await client.get_settings()
+        for entry in all_settings["scb:network"]:
+            if "Hostname" in entry.id:
+                hostname_id = entry.id
+                break
```
Duplicated code. Maybe refactor into a helper function.

stegm replied:
Missing error handling if no hostname id was found. If this unlikely event happens `hostname_id` is undefined and an exception at line 45 is raised.

stegm commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -32,9 +32,17 @@ async def test_connection(hass: HomeAssistant, data) -> str:
+        # get name of the hostname setting id
+        all_settings = await client.get_settings()
+        for entry in all_settings["scb:network"]:
+            if "Hostname" in entry.id:
```
I am not happy with this "heuristic". I would prefer an explicit check against the known ids "Network:Hostname" and "Hostname".

Maybe something like this:

```python
KNOWN_HOSTNAME_IDS = ("Network:Hostname", "Hostname")

def get_hostname_id(client: ApiClient) -> str:
    all_settings = await client.get_settings()
    for entry in all_settings["scb:network"]:
        if entry.id in KNOWN_HOSTNAME_IDS:
            return entry.id
    throw Some Exception
```

stegm reviewed (changes requested):
Please also consider adding or extending unit tests.

See `tests/components/kostal_plenticore/test_config_flow.py` for a start.

erikbadman commented:
I have tried to update the PR as per your comments except for the unit tests.

erikbadman committed:
Add try except in get_hostname_id function

erikbadman committed:
Update tests after adding get_hostname_id function

erikbadman commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -407,10 +407,14 @@ def format_em_manager_state(state: str) -> str | None:
+            if entry.id in KNOWN_HOSTNAME_IDS:
+                return entry.id
+    except (TypeError, ValueError):
+        return "Network:Hostname"
```
Hi @stegm I have updated some tests now and they currently pass. I don't think it is correct to return "Network:Hostname" in case it fails but I'm not sure how to get the tests to pass otherwise.
I'm not a developer just trying my best here :)

stegm committed:
Revert "Update tests after adding get_hostname_id function"

This reverts commit 5fa4e533cb18e8d141dbb1b7aed021f80f4312a2.

stegm committed:
Add test for G2 models in config flow.

stegm committed:
Add test for helper module.

stegm committed:
Fix test for numbers.

stegm reviewed (changes requested):
Concerning the tests: I modified the unit tests for you in https://github.com/stegm/erikbadman-core.git branch "add-version-check": may be you could pull these changes into your branch and update this PR accordingly.

erikbadman commented:
> Concerning the tests: I modified the unit tests for you in https://github.com/stegm/erikbadman-core.git branch "add-version-check": may be you could pull these changes into your branch and update this PR accordingly.

Thanks for this, I have pulled them in and updated the PR

erikbadman commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -403,3 +405,16 @@ def format_em_manager_state(state: str) -> str | None:
+        for entry in all_settings["scb:network"]:
+            if entry.id in KNOWN_HOSTNAME_IDS:
+                return entry.id
+    except (TypeError, ValueError):
```
I added before you updated the tests since it failed otherwise for me, I have now removed the try except and added a raise error in case it fails

erikbadman commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -23,6 +23,7 @@
 
 _LOGGER = logging.getLogger(__name__)
 _DataT = TypeVar("_DataT")
+KNOWN_HOSTNAME_IDS = ("Network:Hostname", "Hostname")
```
Thanks, I have added this now

erikbadman commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -403,3 +405,16 @@ def format_em_manager_state(state: str) -> str | None:
+    except (TypeError, ValueError):
+        return "Network:Hostname"
+
+    return None
```
I removed the try except and added a `raise ApiException("Hostname identifier not found in KNOWN_HOSTNAME_IDS")` instead. Is that ok?

erikbadman committed:
Revert "Add try except in get_hostname_id function"

This reverts commit 059f5bd9b413ca060300e09a6a1f0dffb4420f56.

erikbadman committed:
Update variable name with known hostname ids to be private

stegm reviewed (approved):
ok for me

stegm commented:
@erikbadman  you still have to address the cla error https://github.com/home-assistant/core/pull/92198#issuecomment-1537190685

erikbadman commented:
@stegm I have already updated the commits with the correct information (git log looks correct at least). Can I trigger the bot manually to check again?

stegm commented:
> I have double checked and it looks correct as far as I can tell. Please let me know if you have any idea on what is still the problem, thanks.

Actually I have no idea. Can you simple remove the label?

erikbadman commented:
Hi @MartinHjelmare this PR is blocked by the cla-error label but I think everything should be ok. I have done a rebase and updated with the correct Author and email but the bot does not seem to recheck this PR or remove the label.

I don't seem to have access to remove the cla-error label myself, could you take a look?

erikbadman commented:
> > I have double checked and it looks correct as far as I can tell. Please let me know if you have any idea on what is still the problem, thanks.
> 
> Actually I have no idea. Can you simple remove the label?

Don't seem to have access to remove labels. I tried to ping Martin Hjelmare in a comment since I saw he renamed the PR, maybe he could help.
